keps/sig-network/4858-ip-cidr-validation/README.md

```diff
@@ -0,0 +1,886 @@
+### Goals
+
+- Update the validation of all IP-valued and CIDR-valued API fields in
+  core (`k8s.io/api`) types to require stricter validation, to avoid
+  potential security problems with ambiguously-parsable IPs/CIDRs.
+
+    - Don't require pre-existing invalid fields to be fixed on
+      unrelated updates. E.g., if an existing Service has `clusterIP:
+      172.030.099.099`, allow a user to change the Service's
+      `selector` without being forced to fix the bad `clusterIP` at
+      the same time.
+
+    - Allow immutable fields to be fixed if they are invalid. E.g., if
+      an existing Service has `clusterIP: 172.030.099.099`, allow
+      changing it to `clusterIP: 172.30.99.99` even though `clusterIP`
+      is normally immutable.
+
+- Update the [CEL IP/CIDR validation helpers] to use the same code as
+  the new core API validation. (The CEL helpers already have the
+  correct semantics; this is just about not having two separate
+  implementations.)
+
+```
+<<[UNRESOLVED transition ]>>
+
+- MAYBE do a few releases where IP/CIDR values that are invalid
+  according to the new rules cause apiserver warnings, before actually
+  changing the validation.
+
+- MAYBE put the new validation behind a `Deprecated` feature gate, so
+  that people can turn off the new behavior temporarily without having
+  to downgrade?
+
+<<[/UNRESOLVED]>>
+
+
+<<[UNRESOLVED ipv6-canonical-form ]>>
+
+- MAYBE add apiserver warnings when people specify IPv6 addresses in
+  non-canonical form. (e.g., `"FC99:0:0::0123"` rather than
+  `"fc99::123"`.)
+
+- MAYBE make a plan to require IPv6 addresses to always be in
+  canonical form in _new_ APIs
+
+    - Forcing all values to be in canonical form means you can
+      compare/sort/uniquify them as strings rather than needing to
+      parse them first.
+
+    - (Only relevant to IPv6 because all of the non-canonical forms
+      for IPv4 addresses are now invalid.)
+
+    - (It doesn't make sense to force new values of existing API
+      fields to be canonical, because the "compare them as strings"
+      thing doesn't work if you have to worry about some objects
+      having legacy non-canonical values.)
+
+<<[/UNRESOLVED]>>
+
+
+<<[UNRESOLVED non-ip-fields ]>>
+
+- MAYBE tighten/make-consistent the validation of other
+  networking-related fields at the same time. In particular, some
+  fields that are supposed to be hostname-valued also accept IPv4 (but
+  not IPv6) addresses, because they just use
+  `utilvalidation.IsDNS1123Subdomain`, which requires basically
+  "alphanumeric separated by dots", which IPv4 addresses are.
+
+<<[/UNRESOLVED]>>
+
+
+<<[UNRESOLVED non-special-ip ]>>
+
+- MAYBE revisit the use of `ValidateNonSpecialIP`: certain kinds of
+  IPs (loopback, multicast, link-local, etc) do not make sense in many
+  contexts, but we only prohibit them in a few places.
+
+<<[/UNRESOLVED]>>
+
+
+<<[UNRESOLVED cli-validation ]>>
+
+- MAYBE change/consistentify the validation of IP/CIDR values in CLI
```

Comment:
usually this are used as parameters that will fail when `binding`

Reply:
No, I meant, right now you can say `kube-proxy --metrics-bind-address 010.000.000.001` and it will accept that without complaint and bind to `10.0.0.1`. If we're going to force people to use correctly-formed IPs in the API, maybe we should also force them to use correctly-formed IPs in CLIs/config files.

(Except that we have to be careful to not immediately break their existing startup scripts on upgrade.)

Reply:
but configurations parameters are consumed by the component only, those IPs will be interpreted only for that component.
AFAIK the problem is when there is no consistency in the interpretations, and in our case is when we publish them in the APIs , that should be used consistently across components, if they different interpretations then you can do malicious things ... but if you are consuming one IP value, meanwhile you are consistent in the interpretation , there should be not problem

Reply:
True. And in cases where the CLI value gets copied into an API object (eg, `kube-apiserver --service-cluster-ip-range` → `ServiceCIDR`) the validation on the API object ensures that the component is either validating or normalizing the CLI value.

But there could be cases where a CLI value gets passed directly to an external component, or copied to something other than an API object (eg, `resolv.conf`). And as long as the component normalizes the value, then that's OK, but it means we need to make sure we're doing the right thing in every place, rather than having validation do that for us...